File: Leetcode/Python/437. Path Sum III/solution.py
# https://leetcode.com/problems/path-sum-iii/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        return self.search(root, targetSum, defaultdict(int), 0)

    def search(self, root: Optional[TreeNode], targetSum: int, pathSumHash: Dict[int, int], totalSum: int):
        if not root:
            return 0

        totalSum += root.val
        key = totalSum - targetSum
        pathCount = pathSumHash[key]

        if totalSum == targetSum:
            pathCount += 1

        self.savePathHash(pathSumHash, totalSum, 1)
        pathCount += self.search(root.left, targetSum, pathSumHash, totalSum)
        pathCount += self.search(root.right, targetSum, pathSumHash, totalSum)
        self.savePathHash(pathSumHash, totalSum, -1)

        return pathCount

    def savePathHash(self, pathSumHash: Dict[int, int], key: int, diff: int):
        pathSumHash[key] += diff
        if pathSumHash[key] == 0:
            del pathSumHash[key]

File: Leetcode/Typescript/437. Path Sum III/solution.ts
// https://leetcode.com/problems/path-sum-iii/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number {
  return search(root, targetSum, new Map<number, number>(), 0);
}

function search(
  root: TreeNode | null,
  targetSum: number,
  pathSumHash: Map<number, number>,
  totalSum: number
): number {
  if (!root) {
    return 0;
  }
  totalSum += root.val;
  const key = totalSum - targetSum;
  let pathCount = pathSumHash.get(key) || 0;

  if (totalSum === targetSum) {
    pathCount += 1;
  }

  savePathCount(pathSumHash, totalSum, 1);
  pathCount += search(root.left, targetSum, pathSumHash, totalSum);
  pathCount += search(root.right, targetSum, pathSumHash, totalSum);
  savePathCount(pathSumHash, totalSum, -1);

  return pathCount;
}

function savePathCount(hash: Map<number, number>, key: number, diff: number) {
  const newPathCount = (hash.get(key) || 0) + diff;

  if (newPathCount === 0) {
    hash.delete(key);
  } else {
    hash.set(key, newPathCount);
  }
}
